File: app/src/main/java/com/example/cobe/cryptonews/ui/articles/ArticleAdapter.java
package com.example.cobe.cryptonews.ui.articles;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.cobe.cryptonews.R;
import com.example.cobe.cryptonews.listeners.OnArticleClickListener;
import com.example.cobe.cryptonews.model.Article;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by cobe on 30/03/2018.
 */

public class ArticleAdapter extends RecyclerView.Adapter<ArticleViewHolder> {

    private final List<Article> articles = new ArrayList<>();

    private OnArticleClickListener onArticleClickListener;

    public void setOnArticleClickListener(OnArticleClickListener onArticleClickListener) {
        this.onArticleClickListener = onArticleClickListener;
    }

    public void setArticles(List<Article> articles) {
        this.articles.clear();
        this.articles.addAll(articles);
        notifyDataSetChanged();
    }

    @Override
    public ArticleViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.article_item, parent, false);
        return new ArticleViewHolder(view, onArticleClickListener);
    }

    @Override
    public void onBindViewHolder(ArticleViewHolder holder, int position) {
        Article article = articles.get(position);
        if (article != null) {
            holder.setArticle(article);
        }
    }

    @Override
    public int getItemCount() {
        return articles.size();
    }
}

File: app/src/main/java/com/example/cobe/cryptonews/ui/articles/ArticleAdapter.kt
package com.example.cobe.cryptonews.ui.articles

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.example.cobe.cryptonews.R
import com.example.cobe.cryptonews.listeners.OnArticleClickListener
import com.example.cobe.cryptonews.model.Article

/**
 * Created by cobe on 13/04/2018.
 */
class ArticleAdapter(private val onArticleClickListener: OnArticleClickListener) : RecyclerView.Adapter<ArticleViewHolder>() {

    //private val articles = mutableListOf()<Article>()
    private var articles: MutableList<Article> = ArrayList()

    fun setArticles(articles: List<Article>) {
        this.articles.clear()
        this.articles.addAll(articles)
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: ArticleViewHolder, position: Int) {
        val article = articles[position]
        holder.setArticle(article)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArticleViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.article_item, parent, false)
        return ArticleViewHolder(view, onArticleClickListener)
    }

    override fun getItemCount() = articles.size
}